Enforce native type declarations

greg0ire opened the pull request:
I tried to avoid changes in `tests`, so as to not unnecessarily widen the
gap between 2.x and 3.x.

greg0ire committed:
Enforce native type declarations

I tried to avoid changes in tests, so as to not unnecessarily widen the
gap between 2.x and 3.x.

derrabus commented on phpcs.xml.dist:
```diff
@@ -44,8 +40,17 @@
         <exclude-pattern>tests/Doctrine/Tests/ORM/Functional/ValueObjectsTest.php</exclude-pattern>
     </rule>
 
+    <rule ref="SlevomatCodingStandard.TypeHints.ParameterTypeHint.MissingAnyTypeHint">
+        <exclude-pattern>tests/*</exclude-pattern>
+    </rule>
+
+    <rule ref="SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingNativeTypeHint">
+        <exclude-pattern>tests/*</exclude-pattern>
+    </rule>
```
We should enforce those rules on our tests as well. Excluding (selected) fixtures, maybe.

greg0ire replied:
It will make the merges up harder, and can be done after 3.0.0 is released. I don't remember what exactly what the plan is regarding the merges up and the release of 3.0.0, but I guess the merges up will stop or at least slow down at some point. I was planning on removing these rules then (they result in a huge diff).

derrabus replied:
Sure, that was more like a future TODO. 👍

greg0ire replied:
Oh right, OK 👍

greg0ire committed:
Use native type hint for ProxyFactory::getProxy

This requires heavily adapting tests, because the proxy instance must:
- be an instance of InternalProxy (easy)
- be a valid entity (hard, especially for PHPUnit)

greg0ire commented on tests/Doctrine/Tests/ORM/Hydration/ObjectHydratorTest.php:
```diff
@@ -1021,18 +1024,21 @@ public function testCreatesProxyForLazyLoadingWithForeignKeys(): void
+
+                return parent::getProxy($className, $identifier);
+            }
+        };
```
@derrabus I pushed a new commit, it reworks this test and the one below, allowing to use a native return type declaration for `ProxyFactory::getProxy()`

SenseException reviewed (approved):
I flew over the changes and they seem to fit. :+1:
